src/main/java/org/jitsi/videobridge/VideoConstraintsCompatibility.java

```diff
@@ -92,10 +92,41 @@
     {
         int maxFrameHeightCopy = maxFrameHeight;
 
-        final VideoConstraints selectedEndpointConstraints
-            = new VideoConstraints(Math.min(1080, maxFrameHeightCopy),
-            BitrateControllerConfig.Config.onstagePreferredHeightPx(),
-            BitrateControllerConfig.Config.onstagePreferredFramerate());
+        final VideoConstraints selectedEndpointConstraints;
+        if (selectedEndpoints.size() > 1)
+        {
+            // This implements special handling for tile-view. We can show that
+            // (selectedEndpoints.size() > 1) is equivalent to tile-view (so we
+            // can use it as a clue to detect tile-view):
+            //
+            // (selectedEndpoints.size() > 1) implies tile-view because multiple
+            // "selected" endpoints has only ever been used for tile-view.
+            //
+            // tile-view implies (selectedEndpoints.size() > 1), or, equivalently,
+            // (selectedEndpoints.size() <= 1) implies non tile-view because as
+            // soon as we click on a participant we exit tile-view, see:
+            //
+            // https://github.com/jitsi/jitsi-meet/commit/ebcde745ef34bd3d45a2d884825fdc48cfa16839
+            // https://github.com/jitsi/jitsi-meet/commit/4cea7018f536891b028784e7495f71fc99fc18a0
+            // https://github.com/jitsi/jitsi-meet/commit/29bc18df01c82cefbbc7b78f5aef7b97c2dee0e4
+            // https://github.com/jitsi/jitsi-meet/commit/e63cd8c81bceb9763e4d57be5f2262c6347afc23
+            //
+            // This means that the condition selectedEndpoints.size() > 1 is
+            // equivalent to tile-view.
+
+            // In tile view we set the ideal height but not the preferred height
+            // nor the preferred frame-rate because we want even even
+            // distribution of bandwidth among all the tiles to avoid ninjas.
+            selectedEndpointConstraints
```

Comment:
Maybe not now, but I think in the future we should consider a 'receiver quality policy' to handle this: basically we'd have a policy to prefer quality or number of participants viewed.

Preferring quality would mean that we go through the list of endpoints and pick the maximum quality that can fit for each one until we run out of 'room'

Preferring number of participants viewed would iterate over all the endpoints and get them in their lowest quality, then iterate again to try and improve quality for each one until we ran out of room.

'Stage view' would be the first policy (maximize quality of active speaker), 'tile view' would be the second (maximize the number of participants seen).

Reply:
Yes, the receiver quality policy idea makes sense.

Reply:
Doesn't the "preferred" constraint per source handle this?  If you want to prefer number of participants, set the preferred quality per-source to something low.

Reply:
Τhanks for pointing that out @JonathanLennox, that's exactly the intention

Reply:
> Doesn't the "preferred" constraint per source handle this? If you want to prefer number of participants, set the preferred quality per-source to something low.

I'm not sure I see how it does. How does a receiver know what to set the preferred resolution to?  What if it sets 180, but it actually could have received 360 for all (or some) of the senders?

Reply:
"Preferred" is not "ideal" - it's the minimum to give this source before moving on to the next source.  If every source has reached its "preferred", then the server starts upping resolutions (round-robin, starting with the first) up to "ideal".

Perhaps this is badly named - better names suggested?

Reply:
If that's true then I do think we should come up with a better name.  Maybe `minimum`, as that seems to be what it is.  

But I still think the receiver is not in a good position to set this constraint.  Consider the 'main stage' case: a receiver would set `ideal` and `minimum` to 720p, I assume? What if they don't have the bandwidth to fit that stream?  If the JVB strictly obeys, then it would forward nothing.  It seems like letting the receiver set a policy as to how the JVB would fallback would make it easier for the receiver to express what it wants without code having to 'guess' (either on the receiver side or the JVB's side), then it can express an 'ideal' but let the JVB fit things in based on the available bandwidth.

Reply:
It's not 'minimum' either - the bridge will give less than this value, but if so, won't allocate any bandwidth to any later sources.

To give an example, right now, in main stage the bridge implicitly sets this value to 360p/30 for the selected participant, and 180p/30 for all others.  So it allocates the selected source 360p/30, then allocates bandwidth for the filmstrip sources, and then if there's still available bandwidth ups the main stage to 720p/30.

Reply:
I'd say this other field feels pretty awkward to describe then.  If there's some kind of over-arching 'receiver policy', then a client need only give its ideal height for each sender (or 'pseudo sender', e.g. 'current active speaker') and the policy for how the bridge should degrade if it can't fit that ideal.

1. Stage view: receiver sets ideal to 720p for active speaker, 180p for everyone else.  Policy is "prefer quality": bridge would sort by ideal resolution for senders and fulfill each one completely before moving to the next.

2. Tile view: receiver sets ideal to 720p for all speakers (say there are only 4, so enough screen to fit them in 720p), policy is "prefer number of video streams": bridge iterates through all streams and fits in one layer at a time, going back over them and upgrading the layers while there's still room.

There are some slight twists there with how we decide what order to go over the streams in, but I think that exists in any implementation.